### DISCUSSION TITLE
Avalonia App with Plugin System

### DISCUSSION THREAD

**Discussion by sandreas** [AvaloniaUI/Avalonia#6824] [Q&A]
Hello,

recently I tried Avalonia and it is awesome. My development process is mostly driven by building small personal utilities required to be cross plattform. Now that I have worked out my personal Boilerplate-Project for cross platform deployment (inspired by https://github.com/IngvarX/Camelot), I asked myself if it would be possible to add a small Plugin System for rendering external "classlib" Assemblies into my App-Container UI.

My thoughts:
- Develop an App-Container-Project just holding the MainWindow, a small navigation where Plugins can register to, some basic services (e.g. a settings storage in sqlite, notifications, MessageBus for inter Plugin communication) and the setup for cross platform deployment 
- If I need a new personal tool, I could develop a simple Plugin as classlib without having to create the whole deployment boilerplate, but just putting an Avalonia Plugin DLL into a predefined folder
- (optional) Develop an "App-Store" Plugin able to search for nuget packages of a specific predefined type (e.g. AvaloniaStorePlugin) and manage the installation, so everyone could publish nuget packages using this plugin architecture

I think that this might be an interesting project to push Avalonia in general, if a developer could deploy his personal app just by creating a classlib and then using a container project including only his personal DLL without having to take care of the cross platform deployment stuff.

Would this be possible?
If so, how and does someone have example code, tutorials or hints, to point me into the right direction?

Best regards,
sandreas

**Comment by FSharpCSharp** [comment]
Interesting idea. I am also looking for a similar system. It should not only be easily extensible, but also modular and flexible if possible. Parts of what you're looking for could theoretically be handled by Microsoft's Prism framework. In WPF times this was a tool that covered exactly these points and stood for easy extensibility. Unfortunately, there is only a community-driven development based on Avalonia, which also offers only limited functionality. If there was something like this, then you could build applications much faster, and not have to develop everything from 0 over and over again. It would be really desirable that something like this is available.

**Reply by sandreas** [reply]
Well, seems like there IS a [Prism derivate for Avalonia](https://github.com/AvaloniaCommunity/Prism.Avalonia)... I'll take a look but having an example project with real world code instead of beginner examples would be easier to understand.

**Reply by npolyak** [reply]
In my experience Prism/MEF are too bulky, has a lot of functionality that is not needed and not well documented. Though one can definitely create plugin solutions based on Prism or MEF - I did it myself many times. I plan to create a [UniDock](https://www.codeproject.com/Articles/5316702/UniDock-A-New-Multiplatform-Docking-Framework-UniD) centric plugin system myself, but it will take a bit of time.

**Reply by FSharpCSharp** [reply]
I have (unfortunately) also had to find that out. The project sounds very interesting. Thanks for the tip!

How do you stand here then to the topic MVVM? When examining your interesting articles and examples, I have unfortunately not seen too much about it. You only touch on the subject, and I don't find too much about it in the examples (yet). Does MVVM not play a big role anymore, or did I miss something?

**Reply by npolyak** [reply]
Take a look at [IoCy Container](https://www.codeproject.com/Articles/1244087/Introducing-IoCy-Simple-and-Powerful-IoC-Container). I built it myself and plan to work further on it and keep improving it. If you are interested and notice some important features missing, please tell me, I'll try to add them.

**Reply by npolyak** [reply]
> How do you stand here then to the topic MVVM? When examining your interesting articles and examples, I have unfortunately not seen too much about it. You only touch on the subject, and I don't find too much about it in the examples (yet). Does MVVM not play a big role anymore, or did I miss something?

Hey F#C#, 
sorry I did not notice the second part of your question. Yes, MVVM is very important and I think it will always be. It allows splitting the non-visual functionality from the visual functionality which makes it easier to extend, modify and test. My latest [UniDock](https://www.codeproject.com/Articles/5316702/UniDock-A-New-Multiplatform-Docking-Framework-UniD) article has last 3 samples dedicated to the View Models. 
On the other hand, pure visual controls/framework do not have to be dependent on MVVM - it would only bring an extra layer of complexity. View Models usually serve for wrapping the business logic - adapting the business logic for visual representation/manipulation. 
One way of creating a plugin framework based on the docking functionality is to provide a View Model/View pair for each dock pane's content. Such pair can be developed and tested outside of the plugin framework, but then inserted into the framework almost without any extra effort. Then the user or super-user can configure and save the application's layout containing the individual visual plugins in each of its panes. This is why UniDock allows to specify a view model and a DataTemplate's key to provide the content of a dock pane (DataTemplate can play a role of a view).
Regards

**Reply by FSharpCSharp** [reply]
Thank you for the detailed answer. I'm really excited about your solutions in this area. From my personal point of view, it's really hard to decide which horse to back. There are now many UI libraries around .NET and .NET Core. Avalonia seemed like a great approach because of the platform independence. But unlike other frameworks, such as QT or similar in the C++ world, it simply lacks basic functionality.

I'm glad enough that the developers have at least solved many of the bugs and problems we had in WPF, allowing for a significant speedup in development.

I'll keep watching your developments and experiment with them.

**Reply by npolyak** [reply]
> Avalonia seemed like a great approach because of the platform independence. But unlike other frameworks, such as QT or similar in the C++ world, it simply lacks basic functionality.

What do you think is missing in Avalonia? I thought the package was quite complete at least for Desktop development (aside maybe from a little bit of window manipulation functionality). 

The problem with C++ is that the compilation time is prohibitive - it considerably slows down the prototyping and the development experience. And it is not only because of compiling to the machine code, but also because the Templates in C++ are made in such a way that require code generation at the compilation stage.

**Comment by maxkatz6** [comment]
I might be wrong, but how really plugin system is directly related to the GUI framework? In my opinion it should be Avalonia-depended nor WPF-depended, but abstract .NET implementation with some extensions per specific framework. Is there any example of these abstract solutions?
ReactiveUI and Prism have some of these features, mostly related to navigation, but they don't really provide plugins support.

P.S. In my experience I have wrote an app with plugins only once and it was oversimplified solution. Also from security concerns I always was a fan of solutions like [Windows App Extensions](https://docs.microsoft.com/en-us/windows/uwp/launch-resume/how-to-create-an-extension). which has nothing to do with UI and is strictly isolated.

**Reply by npolyak** [reply]
Often you want different **Visual** plugins to be in different docking panes - at least this is how it was done in many of the projects that I was part of.

**Comment by sandreas** [comment]
Here is some more input:

- Avalonia Docking System: https://github.com/wieslawsoltes/Dock
- Prism [Modules documentation](https://prismlibrary.com/docs/modules.html)  and [Examples](https://github.com/PrismLibrary/Prism-Samples-Wpf/tree/master/07-Modules%20-%20Directory)

Seems that Prism is what I've been looking for - but I did not try it out yet. The Docking-System could be used to add like a Dashboard-Behaviour-Thing, but this would not work as well for mobile devices in future releases of Avalonia supporting iOS and Android, so maybe this would be the wrong direction.

A very basic feature set would contain:
- Inter-Plugin-Communication (e.g. MQTT like)
- Settings / Options provider and storage
- Translations and locales
- Custom menu items
- Maybe Database
- Maybe Fulltext Indexing (e.g. for Documents and Files - like an email Client)


I would like to keep it simple and flexible for the moment, but I also thought about having an App Store within the Project that [connects to the nuget repositories](https://docs.microsoft.com/en-us/nuget/reference/nuget-client-sdk) to install new Plugins / Modules. My worktitle for the project is Pomorum - pretty likely it will never be released, but at least I now do have a starting point :-)

**Reply by FSharpCSharp** [reply]
Glad to hear that Prism seems to basically meet your expectations. But from my point of view there are some current problems in connection with Avalonia:

1. there is only a community-driven version of it and not an official one.
2. due to point 1, it is currently lagging far behind the official Prism state from the feature side. The latest version is still based on version 7.X of the framework, although version 8.X is already available. There is a branch and a roadmap to connect it to the 8.X version, but the implementation is very slow and hesitant.
3. many functions do not seem to work at all. This includes, for example, the navigation journal.
4. a DialogService, which plays a big role in Prism, is currently not really supported.

If you want to develop your framework based on this, you currently have to come up with solutions for these issues yourself. So far, this has kept me from being able to really use it for large apps in a meaningful way.

But as you can see, there are currently a lot of good approaches, but there is always a small but consistently important particle missing to be able to use it meaningfully.

From my point of view, what also plays an important role in today's world is the license. If you build a framework, then it would make sense to use licenses like the MIT license. There are currently a lot of great approaches, but I can't use many of them, because they partly use the GPL license. This makes using it in a large piece of proprietary software impossible for me.

**Comment by Onebeld** [comment]
You could probably use it.

I tried to implement a plugin system (modules) in the [Regul](https://github.com/Onebeld/Regul) repository (I didn't know Prism existed then). I wanted to show it here once, but I forgot. I had to take an off-the-shelf repository ([DotNetCorePlugins](https://github.com/natemcmaster/DotNetCorePlugins)) to load DLL files and modify it slightly to work on .NET Standard. After that:

- I can write plugins in .NET Standard, that is, the plugin will work in the application on both .NET Framework and .NET (Core) 3+;
- Plugins will be cross-platform (currently not implemented in the repository);
- We can safely implement the GUI in plugins and use it in the program;
- Own localization support.

There is still a lot of work, but you can already create plugins and use them.

**Reply by FSharpCSharp** [reply]
This looks great indeed, and offers a lot of great possibilities. The only thing that currently gives me a stomach ache is the license. Unlike Prism, which is licensed under MIT, it is currently difficult or impossible to use in proprietary solutions because of the GPL license. This is a blessing and a curse at the same time. If the license were different, then this framework would indeed be a gamechanger.

**Reply by Onebeld** [reply]
> This looks great indeed, and offers a lot of great possibilities. The only thing that currently gives me a stomach ache is the license. Unlike Prism, which is licensed under MIT, it is currently difficult or impossible to use in proprietary solutions because of the GPL license. This is a blessing and a curse at the same time. If the license were different, then this framework would indeed be a gamechanger.

In the distant future, I will definitely move this functionality to a separate repository with an MIT license; it should exist for everyone.

**Comment by FSharpCSharp** [comment]
I found, rather by accident, the following demo application built entirely with Avalonia.

https://github.com/wieslawsoltes/Core2D

From my point of view a great example, because it contains almost everything that is needed from my point of view. On the one hand a "dock", various tools, a classic designer etc..

Also I find the whole infrastructure very good. There you can indeed learn a lot and draw your conclusions.

Since I find many of the points there very well implemented and developed in terms of dynamics and abstractness and I see these in many applications again, I would find it great if you could perhaps also publish more components as a kind of basic building blocks.

Therefore I mark here now times directly @wieslawsoltes 

It would be nice if you could perhaps briefly comment on this. Would that be conceivable for you to provide the basic components (which everyone finds new again and again) also as separate components, similar to the docking system?

**Comment by wieslawsoltes** [comment]
> I found, rather by accident, the following demo application built entirely with Avalonia.
> 
> [wieslawsoltes/Core2D](https://github.com/wieslawsoltes/Core2D)
> 
> From my point of view a great example, because it contains almost everything that is needed from my point of view. On the one hand a "dock", various tools, a classic designer etc..
> 
> Also I find the whole infrastructure very good. There you can indeed learn a lot and draw your conclusions.
> 
> Since I find many of the points there very well implemented and developed in terms of dynamics and abstractness and I see these in many applications again, I would find it great if you could perhaps also publish more components as a kind of basic building blocks.
> 
> Therefore I mark here now times directly @wieslawsoltes
> 
> It would be nice if you could perhaps briefly comment on this. Would that be conceivable for you to provide the basic components (which everyone finds new again and again) also as separate components, similar to the docking system?

@FSharpCSharp 

I think all of the stuff that Core2D uses is available as components via NuGet
- https://github.com/wieslawsoltes/Dock - docking system
- https://github.com/wieslawsoltes/AvaloniaBehaviors - behaviors including drag and drop helpers
- https://github.com/wieslawsoltes/PanAndZoom - pan and zoom control
- https://github.com/wieslawsoltes/ThemeEditor - color picker control
- https://github.com/wieslawsoltes/Svg.Skia - svg control, svg import
- https://github.com/wieslawsoltes/SvgToXaml - svg to avalonia xaml
- https://github.com/wieslawsoltes/Core2D - even drawing control is now available on NuGet split into reusable packages - [sample](https://github.com/wieslawsoltes/Core2D/tree/master/samples/IntegrationTest) app

also other reusable components for Avalonia all plugable and available from NuGet :
- https://github.com/wieslawsoltes/NodeEditor - easy to use node editor control
- https://github.com/wieslawsoltes/DataBox - DataGrid alternative
- https://github.com/wieslawsoltes/ReactiveHistory - undo/redo framework
- https://github.com/wieslawsoltes/Spreadsheet - excel like control (not yet fully finished, but may be useful)
- https://github.com/wieslawsoltes/Avalonia.ThemeManager - theme switcher
- https://github.com/wieslawsoltes/TypefaceUtil - xaml icon generator from fonts

Also take a look a https://github.com/VitalElement/AvalonStudio.Shell and https://github.com/VitalElement/AvalonStudio

**Reply by cordasfilip** [reply]
@wieslawsoltes Hi I am not sure about this question, but I feel like there is something missing about making a proper plugin system. Is there a way to run the plugin in an isolated mode? So the added plugin is isolated and can't access the shell window and inject unwanted code? Like render a component in a separate process and render the view without having access to the static things like the Application object?

**Reply by wieslawsoltes** [reply]
That isn't part of Avalonia missing but part of .NET you need to implement.

**Reply by FSharpCSharp** [reply]
Thanks a lot for providing all the links. I find your personal achievement regarding the development of the world around AvaloniaUI really great and noteworthy.

But with my original question I didn't want to aim at that. What I personally miss a little bit, and there are for example for MAUI and WPF meanwhile like sand on the sea, is a kind of basic template that you can simply take, and via RAD get faster to the goal. In the above examples of the other UI frameworks, there are already prototypes that include DI, docks, a bootstraper, etc.. These are partly provided by the main developers themselves, or by the community. By not having to reinvent the wheel over and over again, you save a lot of time, especially here. Because mostly the basic requirements are always the same.

But maybe you are even the right person here. Could you perhaps prepare something here that combines these components, a kind of template? You could use as a basis, for example, your example application for the dock developed by you, and further expand DI, so that you have a solid basis, and can start directly.

Of course you could theoretically use the example programs you developed, but I personally always have a bad feeling to simply use the code of others, when you only need a development base. In addition, one is then busy for a while to remove everything not needed, and to reduce to the essential. Your dock is very strong, but due to the fact that I haven't found any documentation (probably due to the time) you have to "bite your way through" to interpret it for yourself, to understand your thinking, and to transfer it into your own applications.

This is not necessary for WPF or MAUI for example. There are very strong and extensive templates with which you often reach your goal very quickly and can really work with RAD principles.

**Reply by cordasfilip** [reply]
> That isn't part of Avalonia missing but part of .NET you need to implement.

Not really what I was getting at. I was thinking more on the lines of iframes in HTML where you can render HTML in an isolated manner. You can run dot net in a new process that will provide the isolation but there isn't anything that can render the combined output to a single screen like with an Iframe.
